### DISCUSSION TITLE
Unique recurring task

### DISCUSSION THREAD

**Discussion by Unkn0wnCat** [hibiken/asynq#376] [Q&A] [answered]
Hi everyone,

for my project I need to have a task run at regular intervals to go through the database and check which entries are due for refresh. I don't want every worker instance to run this, and would like the task to be assigned like any other "normal" task.

For this I need to create a task that may only exist once, but I am stuck on how to facilitate that, as there are N frontend servers and N workers and no single one is supposed to be an authority.

Is this possible? How can I create a task exactly once?

Thanks in advance,
Kevin

**Comment by hibiken** [comment, marked as answer]
@Unkn0wnCat Thank you for opening a discussion!

Have you taken a look at this wiki page on periodic-tasks (https://github.com/hibiken/asynq/wiki/Periodic-Tasks)?

**Update:**
TLDR; If you have multiple `Scheduler`s running to create periodic tasks but want to avoid enqueueing duplicate tasks, try using `TaskID` and `Retention` options. 

```go
// This ensures there's only one task with ID "foo" in the default queue, and the task will be kept in the queue for x duration after the task has been successfully processed.
client.Enqueue(task, asynq.TaskID("foo"), asynq.Retention(x))
```

Also, check issue  #369 to see if there's alternative solution available (e.g. We may decide to add another variant of `Unique` option to keep the uniqueness lock until TTL expires, regardless of the task being processed or not).

**Reply by Unkn0wnCat** [reply]
Hi @hibiken, Thanks for your quick response.

I've looked at the wiki page, but I didn't figure out how to make only one scheduler run at any given time. I tried adding the Unique option as described on [the wiki](https://github.com/hibiken/asynq/wiki/Unique-Tasks#using-unique-option) to no avail.
I've got it to schedule the task, but every server creates the task every minute, and it gets accepted every time. So for 4 servers my "once-per-minute" tasks runs 4 times per minute. Is there a way to prevent that?

**Reply by hibiken** [reply]
I see. Correct me if I'm wrong, but it sounds like you are running the `asynq.Scheduler` together with `asynq.Server` within the same process, and you have four of these processes.

As described in the https://github.com/hibiken/asynq/wiki/Periodic-Tasks, _You have to ensure only a single scheduler is running for a schedule at a time, otherwise you’d end up with duplicate tasks._.

I can think of a few solutions:
1) Run a separate process for `asynq.Scheduler`, and ensure only one such process is running.
2) Add a flag to specify which one of four process is responsible for scheduling periodic tasks.
3) Use `TaskID` or (`Unique`) option to ensure only one task gets enqueued by multiple `Scheduler`s running concurrently

---
Let me know if you have more questions! I'm happy to help :)

**Reply by Unkn0wnCat** [reply]
My program is split in two parts. One cluster runs the frontend and queue client, and one cluster runs the queue server and works on tasks. In the latter one I started a goroutine before the server setting up the Scheduler with the `Unique`-option set. This however meant there were as many Schedulers as there were server instances. The Unique option didn't seem to help. I tried manually setting up cron and posting the tasks with the `Unique`-option, but it didn't work there either. Maybe 1 minute timeout is too short?

Edit: To touch on the only one process handling the scheduling - I really don't want to do that, as I'd like a system where any single instance / server can fail for whatever reason, and I want the system to continue running without any errors in that case

**Reply by hibiken** [reply]
I hear your point about redundancy to tolerate failures 👍 

The use of the `Unique` option seems correct and it should be preventing duplicate tasks from being enqueued within the time window you specify via the argument to the `Unique` function. Would you mind providing code snippets so that I can look into this more closely?

**Reply by Unkn0wnCat** [reply]
Thanks for taking the time to help me with this! Maybe I just made a really obvious mistake, but I can't figure it out for the life of me...

Here's the important snippet from my code. This is run using `go setupScheduledTasks()` before setting up the `asynq.Server`

```golang
type cronJob struct {
	client *asynq.Client
}

func (job *cronJob) Run() {
	log.Println("Enqueueing new cron:minute task")

	task := asynq.NewTask(tasks.TypeCronMinute, nil)

	taskInfo, err := job.client.Enqueue(task, asynq.Unique(time.Minute))
	if err != nil {
		if !errors.Is(err, asynq.ErrDuplicateTask) {
			log.Printf("Unexpected error: %v", err)
			return
		}

		log.Println("The new task violates uniqueness requirement")
		return
	}

	log.Printf("cron:minute task enqueued as #%s", taskInfo.ID)
}

func setupScheduledTasks() {
	client := asynq.NewClient(asynq.RedisClientOpt{Addr: viper.GetString("redis")})

	scheduler := cron.New()

	job, err := scheduler.AddJob("@every 1m", &cronJob{client: client})
	if err != nil {
		log.Printf("Failed to set up cron job")
		return
	}

	log.Printf("Cron job successfully set up as %d", job)

	scheduler.Start()
}
```

For now my cron handler is just a placeholder: 
```golang
func HandleCronMinuteTask(_ context.Context, _ *asynq.Task) error {
	log.Println("Hi from CRON")

	return nil
}
```

Example running with 2 workers:
![Worker 1](https://user-images.githubusercontent.com/8211181/148293029-b2c531b6-38e7-4558-ba89-a1fd7ee12f42.png)
![Worker 2](https://user-images.githubusercontent.com/8211181/148293051-bd55ec98-45f2-4243-8163-66e4c6bb3e70.png)

**Reply by hibiken** [reply]
I see. This is due to the semantics of the `Unique` option. As described in the [godoc](https://pkg.go.dev/github.com/hibiken/asynq#Unique), as soon as the task is deleted from the queue (i.e. successfully processed), the uniqueness lock is released so that the same task can be enqueued.

Looking at the logs you shared, the first process enqueues a task at 22:35:36 and that task is immediately processed and deleted from the queue (at 22:35:36). At this point, the uniqueness lock is released so that the second process can enqueue the same task, which it did at 22:35:41.

This exact issue was recently reported #369, it may be worthwhile to revisit the semantics of `Unique` option so that it doesn't release the lock until TTL expires, regardless of whether the task has been processed or not.

In the meantime, you should be able to work around this by passing `Retention(1 * time.Minute)`, that way it will make sure that the first task will stay in the queue after successfully processed and won't be deleted until the uniqueness lock is expired.

```go
client.Enqueue(task, asynq.Unique(1 * time.Minute), asynq.Retention(1 * time.Minute))
```

**Reply by Unkn0wnCat** [reply]
Ah that makes sense, but weirdly enough it still won't work...

I've changed the enqueue line to the following:
```golang
	taskInfo, err := job.client.Enqueue(task, asynq.Unique(1*time.Minute), asynq.Retention(1*time.Minute))
```

Both tasks show up in asynqmon:
![asynqmon screenshot](https://user-images.githubusercontent.com/8211181/148307802-878f69b4-17c2-42db-a3d0-f1bc850f27ac.png)

Instances:
![worker 1](https://user-images.githubusercontent.com/8211181/148307937-68f98a18-00d7-402f-9511-e28c2d986ce3.png)
![worker 2](https://user-images.githubusercontent.com/8211181/148307954-2eff675b-07b1-4535-858e-9e5e4c10120c.png)

I've tried passing a dummy payload as well instead of `nil` but the result was the same:
![asynqmon screenshot](https://user-images.githubusercontent.com/8211181/148308234-a2dcedb2-22fe-4b47-b711-8982c02871c1.png)

**Reply by hibiken** [reply]
Sorry, I was wrong :( Current implementation releases the lock when task gets processed successfully (regardless of whether retention option is provided or not). 

Give this confusion and the issue #369, maybe we should revisit the unique option semantics. We could let the lock expire only after the given TTL expires. Let me reopen the issue and think about this for a bit. 

In the meantime, is there a way to handle duplicate tasks within the handler (i.e. detect duplicate task and ignore).

**Reply by hibiken** [reply]
Following up:
Is there a way to use `TaskID` option instead of `Unique`? Suggested workaround mentioned in  #369 uses the `TaskID` option.

```go
client.Enqueue(task, asynq.TaskID("foo"), asynq.Retention(1 * time.Minute))
```

**Reply by Unkn0wnCat** [reply]
When using `TaskID` and checking for the `asynq.ErrTaskIDConflict` error it works perfectly! Thank you so much for having taken the time to help me with this, and keep up the good work with this amazing package :)

Edit: Can you add this solution to your original response, then I can mark it as the answer and others can use it too?

**Reply by hibiken** [reply]
Updated the first comment. Thanks for providing all the infos to look into this issue together 🙏
